Adding correlation IDs to reconcile loops

arschles opened the pull request:
Signed-off-by: Aaron Schlesinger <aaron@ecomaz.net>

 <!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**

/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind other
-->

**What this PR does / why we need it**:

This PR adds shared correlation IDs to `context.Context`s passed to reconcile loops, so that the API requests made to Azure get logged and tracked properly. That feature is especially useful for debugging.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1310

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add 'x-ms-correlation-id' headers to all Azure API calls via distributed traces.
```

arschles commented:
/kind feature

CecileRobertMichon commented:
/ok-to-test

Welcome @arschles!

devigned reviewed (commented):
WDYT about changing the way that `tele.Tracer.Start()` works such that anytime start is called and the context does not contain a correlationID, one is generated and added to the value bag. This could be done as a best effort kind of thing, so the error could be ignored, or a default value substituted if a new UUID can not be created?

Also, `pkg/trace/corr.go` requires a file header comment. You will see them littered through the other code files. If you run `make test`, it will generate files, lint, and run the local tests.

arschles commented:
@devigned I think that's a great idea! I'm working on both the `tele.Tracer.Start()` change and fixing build issues now.

arschles commented:
/retest

devigned commented on pkg/trace/span.go:
```diff
@@ -0,0 +1,93 @@
+//	}
+//	defer newSpan.End()
+//	doSomething(ctx)
+func StartSpan(
```
WDYT of something like this:

In `./util/tele.go`, we wrap the `otel.Tracer` with our own struct that implements `trace.Tracer`. We then write our own `Start(...)` func which would then delegate back to `otel.Tracer().Start(...)`. That way the interface can stay the same.
```go
package tele

type tracer struct {
    otel.Tracer
}

func (t *tracer) Start(ctx, opName string, opt ...otel.SpanOption) (context.Context, otel.Span) {
    // add correlationID to ctx and continue as normal
    // return the inner implementation of Start
    return t.Tracer.Start(ctx, opName, opt...)
}

```

Then either in the same package or elsewhere, add context helpers to set / fetch the correlationID.

```go
package corr  // or something else...

// FromCtx or some better name you think of
func FromCtx(ctx context.Context) (ctx, string) {
    // create new context.WithValue containing correlationID and return both if it does not exist, else return context and correlationID
}

```

arschles replied:
Sure, happy to do that! A few questions before I get started, just to make sure I'm on the same page

- I'm assuming that you'd want to still call `tele.Tracer().Start(...)` in all the reconcilers, correct?
- Are you ok with creating our own tracer instance (the wrapper you mentioned) and then passing that -- instead of the default one -- to `otel.SetTracerProvider` in `main.go`?
- Do we need `FromCtx` to be exported at all? It seems like it would only be used inside the same package as the custom tracer, but not sure if you're thinking something more

devigned replied:
> I'm assuming that you'd want to still call tele.Tracer().Start(...) in all the reconcilers, correct?

If possible, that would be great.

> Are you ok with creating our own tracer instance (the wrapper you mentioned) and then passing that -- instead of the default one -- to otel.SetTracerProvider in main.go?

:100: 

> Do we need FromCtx to be exported at all? It seems like it would only be used inside the same package as the custom tracer, but not sure if you're thinking something more

I was adding that because I saw you were returning the correlationID in your `StartSpan` func. This led me to think you wanted a way to access it from the context. I think one might need to access the correlationID from the context to be able to add that as a HTTP header for the Azure SDK for Go clients.

arschles replied:
perfect - I'll make these changes.

regarding `FromCtx` - just in case someone wants to access the correlation ID, I'll make `FromCtx` exported like you said.

arschles replied:
@devigned just an update, so you know I haven't forgotten about this! I'm looking through the implementation & docs trying to figure out a way to create a new [`TracerProvider`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TracerProvider) that we can pass to the `SetTracerProvider` function in `main.go`. That would need to create a new `Tracer` which wraps the implementation you referenced above. I think that's going to require a lot of plumbing, which we could make unnecessary if we called a custom function to get either:

- get a new tracer that which wraps the underlying tracer implementation, like what you showed here
- create a new span, which is what's already in this PR

I'll keep digging, but let me know about going with either of the above options in case there isn't a good way to wrap everything we need

arschles replied:
Another option would be to create a custom `TracerProvider` that wraps the jaeger one. Let me know your thoughts on that too?

devigned replied:
Sorry for the delay, @arschles. I was thinking something more like this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/master...devigned:corr

The above diff should add the correlationID to the context without needing to manipulate the signature of the Tracer. wdyt?

arschles replied:
Hi @devigned I see what you mean now! Thanks, I'll implement that.

arschles commented:
@devigned I made approximately the change you suggested. the only material difference is that I made the custom `tracer`'s `Start(...)` method create the new `context.Context` with the correlation ID in it, and also call the wrapped `trace.Tracer.Start()` function. let me know if this is more along the same lines as what you're thinking.

devigned commented on config/rbac/role.yaml:
```diff
@@ -134,8 +134,6 @@ rules:
   - azuremachinepoolmachines/status
   verbs:
   - get
-  - patch
```
I don't think you meant to do this.

arschles replied:
wowza, the generators might have removed this. sorry about that. re-added in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1460/commits/65ca5042956a85e50911c6a7f22d2975480921f6

devigned commented on exp/controllers/azuremachinepool_controller.go:
```diff
@@ -158,19 +157,16 @@ func (ampr *AzureMachinePoolReconciler) SetupWithManager(ctx context.Context, mg
 
 // Reconcile idempotently gets, creates, and updates a machine pool.
 func (ampr *AzureMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx, span := tele.Tracer().Start(
```
I'd hate to lose the trace attributes that were added to this span. Please consider reverting to the previous Start.

arschles replied:
sorry about that! re-added in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1460/commits/f21e1c8a0c67c16b223d2d3e3ac57d00706a36a1

devigned commented on exp/controllers/azuremanagedcluster_controller.go:
```diff
@@ -100,12 +98,7 @@ func (r *AzureManagedClusterReconciler) Reconcile(ctx context.Context, req ctrl.
 	defer cancel()
 	log := r.Log.WithValues("namespace", req.Namespace, "azureManagedCluster", req.Name)
 
-	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureManagedClusterReconciler.Reconcile",
```
Again, the attributes are valuable. Please consider reverting the change to Start.

arschles replied:
sorry about that! re-added in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1460/commits/f21e1c8a0c67c16b223d2d3e3ac57d00706a36a1

devigned reviewed (changes requested):
I like the work done in the `tele` package to create the tracer adapter and enhancing the `Start` func to do a bit more.

Please consider reverting the changes to files outside of `./util/tele`. Some of the changes to invocations of the `Start` func changed the attributes being supplied or were solely formatting changes.

arschles commented:
/retest

arschles commented:
hey @devigned - would you mind taking another look when you have some time?

devigned commented:
> Please consider reverting the changes to files outside of `./util/tele`. Some of the changes to invocations of the `Start` func changed the attributes being supplied or were solely formatting changes.

I'm sorry for the delay. I was hoping you were going to revert the changes outside of `./util/tele` since the `Start` interface had stayed the same. If you feel strongly about those formatting changes, then I not opposed. I was just thinking it would make the change set tighter and leave less question about the intent of the changes.

This change set adds the correlationID, but it doesn't actually propagate it on the wire. That will need to be added in a future PR, or this one if you want to continue it.

Again, sorry for any confusion. I'm excited to get this pulled in.

CecileRobertMichon commented:
@arschles have you thought about how these correlation IDs would get propagated to the actual Azure API calls? I assume a we'd need to set the `x-ms-client-correlation-id` header on each request made to Azure?

arschles commented:
@devigned got it, I've removed most of the formatting changes made earlier - there are a few still here in which I moved the creation of cancellable contexts below the creation of the spans. that is so the new contexts can be based on the contexts that were created by the new span. I do think it's valuable - but not critical - to do that so the cancellable context has the new correlation IDs in it. let me know what you think of that.

also to address your and @CecileRobertMichon's point about percolating these correlation IDs up to the Azure API - I had originally planned on doing that in a follow-up PR when this PR was large, to draw a clear line between the change to add the IDs and the change to send them up to Azure. now that this is smaller, would you prefer I do that all in this PR?

I'm happy to do it either way, whatever makes it easier for you all

CecileRobertMichon commented on exp/controllers/azuremachinepoolmachine_controller.go:
```diff
@@ -127,17 +127,17 @@ func (ampmr *AzureMachinePoolMachineController) SetupWithManager(ctx context.Con
-	logger := ampmr.Log.WithValues("namespace", req.Namespace, "azureMachinePoolMachine", req.Name)
-
-	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureMachinePoolMachineController.Reconcile",
+	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureMachinePoolController.Reconcile",
```
```suggestion
	ctx, span := tele.Tracer().Start(ctx, "controllers. AzureMachinePoolMachineController.Reconcile",
```

arschles replied:
Thanks @CecileRobertMichon !

CecileRobertMichon commented:
> I had originally planned on doing that in a follow-up PR when this PR was large, to draw a clear line between the change to add the IDs and the change to send them up to Azure. now that this is smaller, would you prefer I do that all in this PR?

separate PR is fine with me, up to you

devigned reviewed (approved):
/lgtm
/assign @CecileRobertMichon

arschles commented:
Thanks @CecileRobertMichon - I'll submit a follow-up PR with those changes

devigned commented:
/lgtm

CecileRobertMichon commented on exp/controllers/azuremachinepool_controller.go:
```diff
@@ -158,18 +158,18 @@ func (ampr *AzureMachinePoolReconciler) SetupWithManager(ctx context.Context, mg
+		),
+	)
 	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(ampr.ReconcileTimeout))
```
why the change in order here but not in every controller? azuremachine and azurecluster still have the reverse order

arschles replied:
I moved this context creation down to below the creation of the span so that the timeout context includes the correlation ID value. IIRC the others had this change but I reverted them by accident. Would it make more sense to save all these changes for a separate PR?

CecileRobertMichon replied:
either way, as long as the changes are consistent (make the changes here to all the controllers or none of them)

arschles replied:
ok, going to take this change out then because it's not strictly necessary

arschles commented on controllers/azuremachine_controller.go:
```diff
@@ -137,7 +137,8 @@ func (r *AzureMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request
 			attribute.String("name", req.Name),
 			attribute.String("kind", "AzureMachine"),
-		))
+		),
+	)
```
@CecileRobertMichon are you ok with this the the other similar format fixes?

devigned commented:
/retest

CecileRobertMichon commented:
/lgtm
/approve
/retest

arschles commented:
/test pull-cluster-api-provider-azure-e2e
